apps/meteor/app/federation-v2/server/application/sender/RoomInternalHooksValidator.ts

```diff
@@ -0,0 +1,97 @@
+	) {
+		super(bridge, internalUserAdapter, internalSettingsAdapter);
+	}
+
+	public async canAddFederatedUserToNonFederatedRoom(internalUser: IUser | string, internalRoom: IRoom): Promise<void> {
+		if (isRoomFederated(internalRoom)) {
+			return;
+		}
+
+		if (this.isAddingANewExternalUser(internalUser)) {
+			throw new Error('error-cant-add-federated-users');
+		}
+
+		const user = await this.internalUserAdapter.getFederatedUserByInternalId((internalUser as IUser)._id);
+		const isAFederatedUser = user?.isRemote();
+		if (isAFederatedUser) {
+			throw new Error('error-cant-add-federated-users');
+		}
+	}
+
+	public async canAddFederatedUserToFederatedRoom(
+		internalUser: IUser | string,
+		internalInviter: IUser,
+		internalRoom: IRoom,
+	): Promise<void> {
+		if (!isRoomFederated(internalRoom)) {
+			return;
+		}
+		if (this.isAddingANewExternalUser(internalUser) && !isDirectMessageRoom(internalRoom)) {
+			throw new Error('error-this-is-an-ee-feature');
+		}
+
+		const inviter = await this.internalUserAdapter.getFederatedUserByInternalId(internalInviter._id);
+		const externalRoom = await this.internalRoomAdapter.getFederatedRoomByInternalId(internalRoom._id);
+		if (!externalRoom || !inviter) {
+			return;
+		}
+
+		const isRoomFromTheProxyServer = FederatedRoom.isOriginalFromTheProxyServer(
+			this.bridge.extractHomeserverOrigin(externalRoom.getExternalId()),
+			this.internalHomeServerDomain,
+		);
+		const isInviterFromTheProxyServer = FederatedUser.isOriginalFromTheProxyServer(
+			this.bridge.extractHomeserverOrigin(inviter.getExternalId()),
+			this.internalHomeServerDomain,
+		);
+		const fullActionExecutedOnTheRemoteHomeServer = !isRoomFromTheProxyServer && !isInviterFromTheProxyServer;
+		if (fullActionExecutedOnTheRemoteHomeServer) {
+			return;
+		}
+
+		const invitee = await this.internalUserAdapter.getFederatedUserByInternalId((internalUser as IUser)._id);
+		const addingAnExternalUser = invitee?.isRemote();
+		const addingExternalUserToNonDirectMessageRoom = addingAnExternalUser && !isDirectMessageRoom(internalRoom);
+		if (addingExternalUserToNonDirectMessageRoom) {
+			throw new Error('error-this-is-an-ee-feature');
+		}
+	}
+
+	public async canCreateDirectMessageFromUI(internalUsers: (IUser | string)[]): Promise<void> {
+		const usernames: string[] = internalUsers.map((user) => {
+			if (this.isAddingANewExternalUser(user)) {
+				return user as string;
+			}
+			return (user as IUser).username || '';
+		});
+		const atLeastOneExternalUser =
+			usernames.some(
+				(username) =>
+					!FederatedUser.isOriginalFromTheProxyServer(this.bridge.extractHomeserverOrigin(username), this.internalHomeServerDomain),
+			) || internalUsers.filter((user) => !this.isAddingANewExternalUser(user)).some((user) => isUserFederated(user as IUser));
+		if (atLeastOneExternalUser) {
+			throw new Error('error-this-is-an-ee-feature');
+		}
+	}
+
+	private isAddingANewExternalUser(user: IUser | string): boolean {
```

Comment:
if you change this function to a type assertion you can remove the type casting all over this file where you need to do like `internalUser as IUser`
```suggestion
	private isAddingANewExternalUser(user: IUser | string): user is string {
```